### PR TITLE
Client Generalization Refactor 

### PR DESCRIPTION
**Pull request by marbar3778**
- [x] Updated the code in accordance with the refactor that has been merged
- [x] Update tutorial to reflect the changes

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,67 +3,21 @@ module github.com/cosmos/sdk-application-tutorial
...
-	google.golang.org/grpc v1.19.1
-	gopkg.in/yaml.v2 v2.2.2
+	golang.org/x/sys v0.0.0-20190329044733-9eb1bfa1ce65 // indirect
+	google.golang.org/genproto v0.0.0-20190327125643-d831d65fe17d // indirect
+	google.golang.org/grpc v1.19.1 // indirect
 )
 
 replace golang.org/x/crypto => github.com/tendermint/crypto v0.0.0-20180820045704-3764759f34a5
+
+replace github.com/cosmos/cosmos-sdk => github.com/cosmos/cosmos-sdk v0.28.2-0.20190605232616-5f9c3fdf8895
```
Comment (marbar3778) [inline go.mod]: will leave this till the next release

File: x/nameservice/alias.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package nameservice
```
Comment (marbar3778) [inline x/nameservice/alias.go]: needed to add this file because import cycles became a problem

**Comment by alexanderbez** [comment]
Is this R4R @marbar3778? I noticed a merge conflict.

File: cmd/nscli/main.go [hunk 3]
```diff
@@ -13,27 +13,10 @@ import (
 	"github.com/spf13/viper"
 	amino "github.com/tendermint/go-amino"
 	"github.com/tendermint/tendermint/libs/cli"
-
```
Comment (okwme) [inline cmd/nscli/main.go]: fails the linter

File: x/nameservice/client/cli/tx.go [hunk 4]
```diff
@@ -6,12 +6,30 @@ import (
 	"github.com/cosmos/cosmos-sdk/client/context"
 	"github.com/cosmos/cosmos-sdk/client/utils"
 	"github.com/cosmos/cosmos-sdk/codec"
-	"github.com/cosmos/sdk-application-tutorial/x/nameservice"
```
Comment (okwme) [inline x/nameservice/client/cli/tx.go]: fails lint

File: tutorial/README.md [hunk 5]
```diff
@@ -60,20 +64,22 @@ Then, just follow along! The first step describes the design of your application
 1. [Design](./app-design.md) the application.
 2. Begin the implementation of your application in [`./app.go`](./app-init.md).
 3. Start building your module by defining some basic [`Types`](types.md).
-4. Create the main core of the module using the [`Keeper`](./keeper.md).
-5. Define state transitions through [`Msgs` and `Handlers`](./msgs-handlers.md).
+4. Define the keys needed for your module [`key`](./key.md)
+5. Create the main core of the module using the [`Keeper`](./keeper.md).
+6. Define state transitions through [`Msgs` and `Handlers`](./msgs-handlers.md).
    - [`SetName`](set-name.md)
```
Comment (okwme) [inline tutorial/README.md]:
```suggestion
    - [`SetName`](./set-name.md)
```
should work anyway but for consistency

File: tutorial/README.md [hunk 6]
```diff
@@ -60,20 +64,22 @@ Then, just follow along! The first step describes the design of your application
...
-12. Import your module and [finish building your application](./app-complete.md)!
-13. Create the [`nsd` and `nscli` entry points](./entrypoint.md) to your application.
-14. Setup [dependency management using `go.mod`](./gomod.md).
-15. [Build and run](./build-run.md) the example.
-16. [Run REST routes](run-rest.md).
+7. Make views on your state machine with [`Queriers`](./queriers.md).
+8. Create the [`alias file`](./alias.md)
+9. Register your types in the encoding format using [`sdk.Codec`](./codec.md).
+10. Create [CLI interactions for your module](./cli.md).
+11. Create [HTTP routes for clients to access your nameservice](rest.md).
```
Comment (okwme) [inline tutorial/README.md]:
```suggestion
11. Create [HTTP routes for clients to access your nameservice](./rest.md).
```

File: tutorial/README.md [hunk 7]
```diff
@@ -60,20 +64,22 @@ Then, just follow along! The first step describes the design of your application
...
+12. Implement the [AppModule interface](module.md)
```
Comment (okwme) [inline tutorial/README.md]:
```suggestion
12. Implement the [AppModule interface](./module.md)
```

File: tutorial/README.md [hunk 8]
```diff
@@ -60,20 +64,22 @@ Then, just follow along! The first step describes the design of your application
...
+13. Configure your [Genesis state](genesis.md).
```
Comment (okwme) [inline tutorial/README.md]:
```suggestion
13. Configure your [Genesis state](./genesis.md).
```

File: tutorial/README.md [hunk 9]
```diff
@@ -60,20 +64,22 @@ Then, just follow along! The first step describes the design of your application
...
+14. Import your module and [finish building your application](./app-complete.md)!
+15. Create the [`nsd` and `nscli` entry points](./entrypoint.md) to your application.
+16. Setup [dependency management using `go.mod`](./gomod.md).
+17. [Build and run](./build-run.md) the example.
+18. [Run REST routes](run-rest.md).
```
Comment (okwme) [inline tutorial/README.md]:
```suggestion
18. [Run REST routes](./run-rest.md).
```

File: tutorial/queriers.md [hunk 10]
```diff
@@ -1,9 +1,42 @@
...
+
+// implement fmt.Stringer
+func (n QueryResNames) String() string {
+	return strings.Join(n[:], "\n")
+}
+```
+
+## Querier
+
+Start by creating the `./x/nameservice/querier.go` file. This is the place to define which queries against application state users will be able to make. Your `nameservice` module will expose three queries:
```
Comment (okwme) [inline tutorial/queriers.md]:
```suggestion
Now you can create the `./x/nameservice/querier.go` file. This is the place to define which queries against application state users will be able to make. Your `nameservice` module will expose three queries:
```

File: go.mod (hunk 1 above)
Comment (alexanderbez) [reply go.mod]: I think we can just update the existing version instead of using a replace (and maybe add a TODO).

**Pull request merged**